lib/grape/middleware/base.rb

```diff
@@ -21,9 +21,17 @@ def call(env)
 
       def call!(env)
         @env = env
+        error_class = nil
         before
-        @app_response = @app.call(@env)
-        after || @app_response
+        begin
+          @app_response = @app.call(@env)
+        rescue => e
+          error_class = e
+        ensure
+          after_response = after
+        end
+
+        after_response || (error_class ? (fail error_class) : @app_response)
       end
```

Comment:
I think we shouldn't be expecting the `after` block to return a response. Otherwise we need to change the behavior of `before` as well and this will definitely have impact on API implementations out there. These are callbacks, so I think they should be called, but their results discarded, at least as the next increment.